feat(instrumentation-aws-sdk): add bedrock extension to apply gen ai conventions

anuraaga opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/guides/contributor#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

- AWS SDK spans for bedrock should be gen AI spans but are currently generic SDK spans

## Short description of the changes

- Add a service extension for bedrock, applying gen AI conventions
  - This initial PR is to get the general infrastructure and setup so applies minimally, only `Converse` with span attributes. Future PRs will add other bedrock APIs and other gen AI conventions such as events and metrics
- Update AWS SDK deps to sync up with the now newer required SDK version

/cc @trentm @codefromthecrypt

anuraaga commented:
Hi @AmanAgarwal041 - just wanted to share this PR adding gen ai instrumentation of bedrock. Notably, it takes a pattern of recording real LLM responses using nock back, which I see most gen ai instrumentation outside of opentelemetry taking due to the complexity of the models. It may be a good approach for #2402 (if you need help with that PR, let me know). Cheers.

anuraaga commented:
@trentm Oh I remembered now that the latest version of the AWS SDK doesn't support Node 14, which is causing the unit tests to fail after https://github.com/anuraaga/opentelemetry-js-contrib/pull/1#discussion_r1942108599. Let me know what's a good way forward, I think either dropping Node 14 support for this package (not sure it's allowed) or downgrading again.

codefromthecrypt reviewed (approved):
Surprisingly concise change to add the toehold here. I know there's more to do, but good work!

jj22ee commented:
Hi @anuraaga, thanks for this. Just to let you know, the ADOT team from AWS also had vested interest in Gen AI support in AWS SDK instrumentation.

There are Bedrock Service Extension implementations in multiple languages in the ADOT's downstream of the auto-instrumentations ([JS](https://github.com/aws-observability/aws-otel-js-instrumentation/blob/v0.5.0/aws-distro-opentelemetry-node-autoinstrumentation/src/patches/aws/services/bedrock.ts#L203), [Python](https://github.com/aws-observability/aws-otel-python-instrumentation/blob/v0.8.0/aws-opentelemetry-distro/src/amazon/opentelemetry/distro/patches/_bedrock_patches.py#L249)), which are similar but might not be fully 1:1 with your changes. I was wondering if you could also consider ADOT's implementations in this PR? I'll also spend time to compare the 2 implementations for any subtle differences.

codefromthecrypt commented:
@jj22ee drive-by comment, but one thing to give heads up about is that the toehold here is based on the latest genai semantic conventions, and I don't expect this to want to do too much in one PR. So, one thing to think about is what happens in this PR vs a follow-up to add more features or things that are defined beyond the otel specs. Doing it like that might result in the same or similar end, but faster vs trying to do too much in the first of many PRs. food for thought!

https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-spans.md

jj22ee commented:
Thanks for the heads up @codefromthecrypt. I agree to start smaller. Just wanted to ensure there are no conflicts, but the latest Gen AI conventions will take priority, so I have no issue with that.

jj22ee commented on plugins/node/opentelemetry-instrumentation-aws-sdk/src/services/bedrock-runtime.ts:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    const { stopReason, usage } = response.data;
+    const { inputTokens, outputTokens } = usage;
+    if (inputTokens !== undefined) {
+      span.setAttribute(ATTR_GEN_AI_USAGE_INPUT_TOKENS, inputTokens);
+    }
+    if (outputTokens !== undefined) {
+      span.setAttribute(ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, outputTokens);
+    }
```
In case `usage` is undefined:
```suggestion
    if (usage) {
      const { inputTokens, outputTokens } = usage;
      if (inputTokens !== undefined) {
        span.setAttribute(ATTR_GEN_AI_USAGE_INPUT_TOKENS, inputTokens);
      }
      if (outputTokens !== undefined) {
        span.setAttribute(ATTR_GEN_AI_USAGE_OUTPUT_TOKENS, outputTokens);
      }
    }
```

jj22ee commented on plugins/node/opentelemetry-instrumentation-aws-sdk/src/semconv.ts:
```diff
@@ -0,0 +1,33 @@
+ * limitations under the License.
+ */
+
+// Gen AI conventions
```
Can you also help copy over comments for the copied `ATTR_*` attribute keys from the [semantic-conventions](https://github.com/open-telemetry/opentelemetry-js/blob/main/semantic-conventions/src/experimental_attributes.ts)?

[Example from a resource-detector package](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/resource-detector-aws-v1.11.0/detectors/node/opentelemetry-resource-detector-aws/src/semconv.ts#L17-L30)

anuraaga replied:
Thanks @trentm - feel free to push anything you need to this branch

karthikscale3 commented:
nice work! looks good for the most part.

anuraaga commented:
Thanks @jj22ee - I think the implementation pattern itself is similar/same as almost all the work is done by the service extension anyways. Technically there isn't any overlap yet since AFAIU, ADOT currently only instruments `InvokeModel` while I started with `Converse` here, though when adding `InvokeModel` support here I think it will be very similar to ADOT. Skimming through the attributes, I didn't see any conflicts in semconv either yet (it's still `gen_ai.system`, for now...)

The bigger difference is the test infrastructure using nock-back, which we also adopted in upstream python

https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-botocore/tests/conftest.py#L71

so I think if we get this in for the initial infrastructure we'll be able to smoothly add more features over subsequent PRs by only focusing on business logic instead of build infra.

jj22ee commented:
Thank you, the changes lgtm!

As for the suddenly failing unit tests, the latest commit shows 15 tests failing, but I'm pretty sure 14 of them are unrelated to your changes. I suppose there is an issue with the recently added `"does not currently add genai conventions"` test which is causing the other 14 to fail. Likely `InvokeModel` isn't intercepted by nock-back?

I recall I had a similar issue when using nock for Kinesis Client, but needed to enforce [NodeHttpHandler requestHandler](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/instrumentation-aws-sdk-v0.49.0/plugins/node/opentelemetry-instrumentation-aws-sdk/test/kinesis.test.ts#L55-L58) since I found that the default handler was HTTP2 which was not supported by nock. Not sure if this issue is the same.

anuraaga commented:
Sorry @jj22ee I missed the test failures behind the known node 14 ones I need advice on. Just forgot to git add the recording will do it soon

trentm commented:
@anuraaga I have *an* idea for dealing with the node v14 test breakage. It may end up looking a little gross. I still have to write it up, and I may break it into a separate PR: a PR that (a) updates the existing `@aws-sdk/client-*` deps to recent versions; (b) changes the `npm test` for this package to just skip running with Node.js v14 and v16; and (c) uses the "test-all-versions" tests (i.e. .tav.yml) to handle running tests with node 14 and 16.

trentm commented:
> I have _an_ idea

#2722 and #2723 for this.

anuraaga commented:
@trentm I have merged main, added bedrock to tav, removed the node `14` and `16` blocks from existing TAV since I guess they're not actually used anymore, and build is green. Hopefully we're ready to go!

jj22ee commented on plugins/node/opentelemetry-instrumentation-aws-sdk/.tav.yml:
```diff
@@ -9,30 +9,27 @@
+"@aws-sdk/client-bedrock-runtime":
   env:
     - SKIP_TEST_IF_DISABLE=true
   # - 3.529.0 was missing the fast-xml-parser dependency (https://github.com/aws/aws-sdk-js-v3/releases/tag/v3.529.1)
```
Can remove this comment since the bedrock-runtime tests are only for `^3.587.0`

jj22ee reviewed (approved):
Adding a note here, in the final followup PR, the README can be updated [here](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-aws-sdk#specific-service-logic) and [here](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-aws-sdk#semantic-conventions) for all the new bedrock service extensions and attributes. I also haven't updated it yet with the s3/kinesis extensions...
